semgrep-core/src/engine/Unit_engine.ml

```diff
@@ -218,15 +218,93 @@ let maturity_tests () =
 (* Language-specific tests *)
 (*****************************************************************************)
 
-let sgrep_file_of_target file =
+let related_file_of_target ~ext ~file =
   let d, b, _e = Common2.dbe_of_filename file in
-  let candidate1 = Common2.filename_of_dbe (d, b, "sgrep") in
-  if Sys.file_exists candidate1 then candidate1
+  let candidate1 = Common2.filename_of_dbe (d, b, ext) in
+  if Sys.file_exists candidate1 then Some candidate1
   else
     let d = Filename.concat tests_path "POLYGLOT" in
-    let candidate2 = Common2.filename_of_dbe (d, b, "sgrep") in
-    if Sys.file_exists candidate2 then candidate2
-    else failwith (spf "could not find sgrep file for %s" file)
+    let candidate2 = Common2.filename_of_dbe (d, b, ext) in
+    if Sys.file_exists candidate2 then Some candidate2 else None
+
+let rec fail_on_overlapping_fixes = function
+  | f1 :: f2 :: tl ->
+      let (_, end1), f1_text = f1 in
+      let (start2, _), f2_text = f2 in
+      if end1 > start2 then
+        failwith
+          (spf "found overlapping fixes:\n\n  %s\n\n  %s" f1_text f2_text);
+      fail_on_overlapping_fixes (f2 :: tl)
+  | [ _ ]
+  | [] ->
+      ()
+
+(* Allows the  semgrep-core test runner that we use to test matches to also test
+ * autofix. The format is pretty simple: add a `.fix` file with the fix pattern
+ * and a `.fixed` file with the expected contents of the target after fixes are
+ * applied.
+ *
+ * There are also end-to-end tests which test autofix on the CLI side.
+ * Unfortunately, modifying them involves updating a large snapshot file, and
+ * maintaining any large quantity of them would be onerous. Additionally, it's
+ * not easy to adapt tests from our large existing test suite in semgrep-core to
+ * also exercise autofix.
+ *
+ * Semgrep's `--test` flag can also test autofix
+ * (https://github.com/returntocorp/semgrep/pull/5190), but it has the same
+ * problems as the existing autofix e2e tests for these purposes. *)
+let compare_fixes lang ~file ~fix matches =
+  match fix with
+  | None -> ()
+  | Some fix ->
+      let expected_fixed_text =
+        let expected_fixed_file =
+          match related_file_of_target ~ext:"fixed" ~file with
+          | Some file -> file
+          | None -> failwith (spf "could not find fixed file for %s" file)
+        in
+        Common.read_file expected_fixed_file
+      in
+      let file_text = Common.read_file file in
+      let fixes =
```

Comment:
This whole part of applying a set of fixes to a file from a list of matches, checking for overlapping,
and apply in order could be put in Autofix.ml.
If Martin's project is successful, we would do those things OCaml side at some point, and
better in Autofix.ml than in testing code.

Reply:
I was planning to productionize this and move it over when I needed it, but I can just put it there now instead with a note that it may not be suitable for use in production.